lib/collection/src/collection.rs

```diff
@@ -1039,22 +1039,120 @@ impl Collection {
         Ok(points)
     }
 
-    #[allow(unreachable_code, clippy::diverging_sub_expression)]
     pub async fn update_params_from_diff(
         &self,
         params_diff: CollectionParamsDiff,
     ) -> CollectionResult<()> {
         let mut config = self.config.write().await;
         config.params = params_diff.update(&config.params)?;
-        self.handle_replica_changes(todo!("supply replica changes"));
+        // TODO supply replica changes
+        let changes = HashSet::new();
+        self.handle_replica_changes(changes).await?;
```

Comment:
I added a `todo` macro here so that when we test, the code will panic if the logic is unimplemented and we will notice it. With a comment it is easier to forget.

Reply:
I had some weird clippy errors with the `todo` macro, I will give it another try!

Reply:
This is the error :sweat_smile: 

```diff
--- a/lib/collection/src/collection.rs
+++ b/lib/collection/src/collection.rs
@@ -1039,15 +1039,14 @@ impl Collection {
         Ok(points)
     }
 
+    #[allow(unreachable_code, clippy::diverging_sub_expression)]
     pub async fn update_params_from_diff(
         &self,
         params_diff: CollectionParamsDiff,
     ) -> CollectionResult<()> {
         let mut config = self.config.write().await;
         config.params = params_diff.update(&config.params)?;
-        // TODO supply replica changes
-        let changes = HashSet::new();
-        self.handle_replica_changes(changes).await?;
+        self.handle_replica_changes(todo!("supply replica changes")).await?;
         Ok(())
     }
```


```
error: future cannot be sent between threads safely
    --> src/tonic/api/collections_api.rs:74:64
     |
74   |       ) -> Result<Response<CollectionOperationResponse>, Status> {
     |  ________________________________________________________________^
75   | |         self.perform_operation(request).await
76   | |     }
     | |_____^ future created by async block is not `Send`
     |
     = help: the trait `Sync` is not implemented for `core::fmt::Opaque`
note: future is not `Send` as this value is used across an await
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:69
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                     ------------------------------- ^^^^^^ await occurs here, with `$crate::format_args!($($arg)+)` maybe used later
     |                                     |
     |                                     has type `ArgumentV1<'_>` which is not `Send`
note: `$crate::format_args!($($arg)+)` is later dropped here
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:76
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                                                            ^
     = note: required for the cast from `impl futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>>` to the object type `dyn futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>> + std::marker::Send`

error: future cannot be sent between threads safely
    --> src/tonic/api/collections_api.rs:81:64
     |
81   |       ) -> Result<Response<CollectionOperationResponse>, Status> {
     |  ________________________________________________________________^
82   | |         self.perform_operation(request).await
83   | |     }
     | |_____^ future created by async block is not `Send`
     |
     = help: the trait `Sync` is not implemented for `core::fmt::Opaque`
note: future is not `Send` as this value is used across an await
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:69
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                     ------------------------------- ^^^^^^ await occurs here, with `$crate::format_args!($($arg)+)` maybe used later
     |                                     |
     |                                     has type `ArgumentV1<'_>` which is not `Send`
note: `$crate::format_args!($($arg)+)` is later dropped here
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:76
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                                                            ^
     = note: required for the cast from `impl futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>>` to the object type `dyn futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>> + std::marker::Send`

error: future cannot be sent between threads safely
    --> src/tonic/api/collections_api.rs:88:64
     |
88   |       ) -> Result<Response<CollectionOperationResponse>, Status> {
     |  ________________________________________________________________^
89   | |         self.perform_operation(request).await
90   | |     }
     | |_____^ future created by async block is not `Send`
     |
     = help: the trait `Sync` is not implemented for `core::fmt::Opaque`
note: future is not `Send` as this value is used across an await
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:69
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                     ------------------------------- ^^^^^^ await occurs here, with `$crate::format_args!($($arg)+)` maybe used later
     |                                     |
     |                                     has type `ArgumentV1<'_>` which is not `Send`
note: `$crate::format_args!($($arg)+)` is later dropped here
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:76
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                                                            ^
     = note: required for the cast from `impl futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>>` to the object type `dyn futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>> + std::marker::Send`

error: future cannot be sent between threads safely
    --> src/tonic/api/collections_api.rs:95:64
     |
95   |       ) -> Result<Response<CollectionOperationResponse>, Status> {
     |  ________________________________________________________________^
96   | |         self.perform_operation(request).await
97   | |     }
     | |_____^ future created by async block is not `Send`
     |
     = help: the trait `Sync` is not implemented for `core::fmt::Opaque`
note: future is not `Send` as this value is used across an await
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:69
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                     ------------------------------- ^^^^^^ await occurs here, with `$crate::format_args!($($arg)+)` maybe used later
     |                                     |
     |                                     has type `ArgumentV1<'_>` which is not `Send`
note: `$crate::format_args!($($arg)+)` is later dropped here
    --> /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:1049:76
     |
1049 |         self.handle_replica_changes(todo!("supply replica changes")).await?;
     |                                                                            ^
     = note: required for the cast from `impl futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>>` to the object type `dyn futures_util::Future<Output = Result<tonic::Response<CollectionOperationResponse>, tonic::Status>> + std::marker::Send`

error: could not compile `qdrant` due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
```

Seeing that it works fine without the `todo!` macro, I got lazy and ditched it without further investigation at the time.

Reply:
Does `panic!` or `unimplemented!` work? Or putting `todo` just as a separate line? Anyway I don't think it is worth much investigation, feel free to just use a comment if it takes too long